Add K8s CiliumEndpoint CRD

raybejjani opened the pull request:
**Summary of changes**
This is very much a WIP, but I wanted to start getting feedback on it.
This adds a new k8s CRD alongside the current CiliumNetworkPolicy. They use the same set/update code and will update in lockstep. The CRD object itself is a copy of the API models.Endpoint struct (actually, it uses it directly) and contains the data in there without modification. `kubectl describe` also works but it is a little hokey.

***Things I'm not sure about***
- Each endpoint has a controller that updates to k8s every second. This involves generating the endpoint model (locking the endpoint) and then posting to k8s.
  - I post to k8s every controller loop. k8s seems intelligent about not updating the data if it hasn't changed. We can do this on the client side, however.
  - Is a 1s interval too short? We can't trigger on a change because our current architecture doesn't really allow for that kind of eventing.

- I'm not sure if I can make these CRDs read-only. I need to add validation logic anyway, but I don't think I can use it to block writes from users. Perhaps some RBAC thing would work.

- I don't do any cleanup of leftover entities on agent start. When an agent crashes, it will leave all CiliumEndpoint objects behind. If it never comes back up (say, the whole node has been removed) nothing will clean it up. In the case of the agent coming back up, we can quite easily run a controller to remove any dangling CRD entries. I don't know how to extend this to cleaning up after crashed-and-never-came-back agents (well, I know how, but it involves complexity).

- I elected to not include the endpoint logs. The most recent log line is included since it's in the endpoint model to begin with.

Fixes: https://github.com/cilium/cilium/issues/2183

```release-note
Cilium exports CiliumEndpoint objects to kubernetes clusters.
```

**How to test (optional)**:
- boot the agent
- install some pods
- 
```
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ kubectl get ciliumendpoints
NAME                          AGE
cilium-k8s-master-cep-11004   57m
cilium-k8s-master-cep-3365    57m
cilium-k8s-master-cep-42910   57m
cilium-k8s-master-cep-49509   57m
```

- 
```json
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ kubectl get ciliumendpoints cilium-k8s-master-cep-11004 -o json
{
    "apiVersion": "cilium.io/v2",
    "endpoint": {
        "addressing": {
            "ipv4": "10.0.54.125",
            "ipv6": "f00d::a0b:0:0:2afc"
        },
        "container-id": "902b1724832d233cc6fbb6ceaf918e4f7576f1220b11c6788b2dfbdbf309fd18",
        "container-name": "k8s_POD_app2_default_62700f83-0db5-11e8-afae-0800271a60aa_0",
        "controllers": [
            {
                "configuration": {
                    "error-retry": true,
                    "interval": "1s"
                },
                "name": "sync-to-k8s-ciliumendpoint (11004)",
                "status": {
                    "failure-count": 3,
                    "last-failure-timestamp": "2018-02-09T08:54:22.669-08:00",
                    "last-success-timestamp": "2018-02-09T09:20:09.006-08:00",
                    "success-count": 1519
                },
                "uuid": "df05a4be-0db9-11e8-b341-0800271a60aa"
            },
            {
                "configuration": {
                    "error-retry": true,
                    "interval": "1m0s"
                },
                "name": "sync-identity-to-k8s-pod (11004)",
                "status": {
                    "last-failure-timestamp": "0001-01-01T00:00:00.000Z",
                    "last-success-timestamp": "2018-02-09T09:19:20.325-08:00",
                    "success-count": 26
                },
                "uuid": "df0a803f-0db9-11e8-b341-0800271a60aa"
            }
        ],
        "health": {
            "bpf": "OK",
            "connected": true,
            "overallHealth": "OK",
            "policy": "OK"
        },
        "host-mac": "0a:8a:e5:98:d1:73",
        "id": 11004,
        "identity": {
            "id": 39681,
            "labels": [
                "k8s:id=app2",
                "k8s:io.kubernetes.pod.namespace=default"
            ],
            "labelsSHA256": "64729f2e0103c24e258d9c9785a5fd87af5d60650d86e01ca9b410badbc24437"
        },
        "interface-index": 4531,
        "interface-name": "lxc902b1",
        "labels": {
            "custom": [],
            "disabled": [],
            "orchestration-identity": [
                "k8s:io.kubernetes.pod.namespace=default",
                "k8s:id=app2"
            ],
            "orchestration-info": [
                "container:io.kubernetes.pod.uid=62700f83-0db5-11e8-afae-0800271a60aa",
                "container:annotation.kubernetes.io/config.seen=2018-02-09T08:22:12.974794312-08:00",
                "container:annotation.kubernetes.io/config.source=api",
                "container:io.kubernetes.container.name=POD",
                "container:io.kubernetes.docker.type=podsandbox",
                "container:io.kubernetes.pod.name=app2"
            ]
        },
        "mac": "ae:3e:27:78:63:75",
        "pod-name": "default:app2",
        "policy": {
            "allowed-consumers": [
                1
            ],
            "build": 5,
            "cidr-policy": {
                "egress": [],
                "ingress": []
            },
            "id": 39681,
            "l4": {
                "egress": [],
                "ingress": []
            }
        },
        "policy-enabled": "none",
        "policy-revision": 5,
        "state": "ready",
        "status": [
            {
                "code": "OK",
                "message": "Successfully regenerated endpoint program due to syncing state to host",
                "state": "ready",
                "timestamp": "2018-02-09T16:54:22Z"
            }
        ]
    },
    "kind": "CiliumEndpoint",
    "metadata": {
        "clusterName": "",
        "creationTimestamp": "2018-02-09T16:22:21Z",
        "generation": 0,
        "initializers": null,
        "name": "cilium-k8s-master-cep-11004",
        "namespace": "default",
        "resourceVersion": "67558",
        "selfLink": "/apis/cilium.io/v2/namespaces/default/ciliumendpoints/cilium-k8s-master-cep-11004",
        "uid": "67b25c4f-0db5-11e8-afae-0800271a60aa"
    }
}
```

raybejjani committed:
k8s: Add CiliumEndpoint CRD

CiliumEndpoint reflects the state of an endpoint. It contains the same
data sent in the cilium endpoint status API.

Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani committed:
k8s: Add CiliumEndpoint sync controller

Each endpoint now runs a controller to update it's state with k8s as a
CiliumEndpoint CRD. CiliumEndpoint reflects the state of an endpoint,
and contains the same data sent in the cilium API. The controller
updates the remote state once a second. On endpoint deletion the
corresponding CRD instance is also deleted.

Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani committed:
endpoint: e.getLogger is nil safe for e.Opts

In some creation paths an Endpoint does not have Opts set when getLogger
is called. It's unfortunate but generally safer to ignore the options
here (mainly Debug).

Signed-off-by: Ray Bejjani <ray@covalent.io>

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -356,14 +362,107 @@ func (e *Endpoint) WaitForProxyCompletions() error {
+		return
+	}
+
+	restConfig, err := k8s.CreateConfig()
```
Isn't there a client in `k8s`? `k8s.Client()`

raybejjani replied:
I tried using this but I couldn't figure out how to get from the return of k8s.GetClient() to a cilium object :( I have a getCiliumClient function now, so I can try something out if you can see the path.

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -51,7 +42,7 @@ const (
 	// CustomResourceDefinitionSchemaVersion is semver-conformant version of CRD schema
 	// Used to determine if CRD needs to be updated in cluster
-	CustomResourceDefinitionSchemaVersion = "1.2"
+	CustomResourceDefinitionSchemaVersion = "1.3"
```
`1.4` (it is funny this won't be catch by git conflict)

raybejjani committed:
k8s: Add CiliumEndpoint sync GC controller

When a node exits, it may leave behind CiliumEndpoint objects. We
remove these, cluster-wide, periodically. We ensure that we don't span
the k8s master by limiting how many cilium agents are likely to execute
a GC on every controller iteration

Signed-off-by: Ray Bejjani <ray@covalent.io>

aanm commented on Makefile:
```diff
@@ -104,7 +104,7 @@ clean-container:
 	-$(MAKE) -C ./contrib/packaging/deb clean
 	-$(MAKE) -C ./contrib/packaging/rpm clean
-	-rm GIT_VERSION
+	-rm -f GIT_VERSION
```
missed signoff

raybejjani replied:
fixed (but not pushed)

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -171,7 +231,9 @@ func createUpdateCRD(clientset apiextensionsclient.Interface, CRDName string, cr
 
 			if clusterCRD.Spec.Validation == nil ||
-				!reflect.DeepEqual(clusterCRD.Spec.Validation.OpenAPIV3Schema, crv.OpenAPIV3Schema) {
+				crd.Spec.Validation == nil ||
```
If you have that comment in line 195, this will always be true

raybejjani replied:
I finally looked at this. It is true but we only get here if the version has changed and needsUpdate returns true. At that point we do want to update it if either is nil in case we add or remove validation at some point.

aanm replied:
Please rebase with master, I've changed this

raybejjani replied:
done

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -207,12 +207,15 @@ func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressin
 		}
+
+		// Set a k8s namespace so a CiliumEndpoint object can be created
+		ep.SetK8sNamespace("kube-system")
```
Transform this into a constant please :-)

raybejjani replied:
good point

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -360,14 +381,125 @@ func (e *Endpoint) WaitForProxyCompletions() error {
+		return
+	}
+
+	ciliumClient, err := getCiliumClient()
```
This will create a client for every `runK8sCiliumEndpointSync` invocation. Can't you share a client for all endpoints?

raybejjani replied:
I should be able to. I'll switch it over.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -360,14 +381,125 @@ func (e *Endpoint) WaitForProxyCompletions() error {
+	}
+
+	var lastMdl *models.Endpoint
+	mdlEqual := func(a, b *models.Endpoint) bool {
```
What is wrong with the good old `reflect.DeepEqual`? Does `MarshalBinary` count for golang out of order maps?

raybejjani replied:
Hehe, I was trying to avoid reflect but I may as well use it for now, until we have something better

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -360,14 +381,125 @@ func (e *Endpoint) WaitForProxyCompletions() error {
+
+				namespace := e.GetK8sNamespace()
+				if err := ciliumClient.CiliumEndpoints(namespace).Delete(epName, &meta_v1.DeleteOptions{}); err != nil {
+					scopedLog.WithError(err).Error("Unable to delete CEP")
```
add the endpoint's name and namespace to the log message

raybejjani replied:
it should already be there, I think. `scopedLog` is derived from the endpoint logger

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -360,14 +381,125 @@ func (e *Endpoint) WaitForProxyCompletions() error {
+
+				_, err = ciliumClient.CiliumEndpoints(namespace).Create(cep)
+				if err != nil {
+					scopedLog.WithError(err).Error("Cannot create CEP")
```
add the endpoint's name and namespace to the log message

raybejjani replied:
This should already be in the `scopedLog` because of `e.getLogger()`

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -985,6 +1121,10 @@ func ParseEndpoint(strEp string) (*Endpoint, error) {
 
+	ep.Mutex.Lock()
+	defer ep.Mutex.Unlock()
+	ep.runK8sCiliumEndpointSync()
```
What's the reason to execute `runK8sCiliumEndpointSync` here?

raybejjani replied:
Isn't this how restored endpoints are built? we don't have a single NewEndpoint function so we need to start the controller in all places we create an endpoint. Should I do it somewhere else for those?

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -166,6 +167,12 @@ func init() {
+
+	// This sleep allows detecting k8s.IsEnabled
+	go func() {
+		time.Sleep(2 * time.Second)
```
Ouch... please tell me there's other way to do this

raybejjani replied:
yeah, I wasn't happy with this. I responded to your comment below. I'll be back :)

raybejjani replied:
I moved this to the k8s_watcher. That way the timing is more correct in relation to having k8s setup. It also means that it won't be started in any program that uses the endpoint package (the cilium-health subprogram was running the controller and logging debug)

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -182,6 +189,74 @@ func getCiliumClient() (cilium_client_v2.CiliumV2Interface, error) {
+	runThrottler := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	// this dummy manager is needed only to add this controller to the global list
+	controller.NewManager().UpdateController(controllerName,
```
What about running the controller with a runinterval of 1 second, and pooling every second for 1 minute to detect if `k8s` is enabled? if it returns false after 60 seconds, we kill the controller

raybejjani replied:
It could work. I should check if we can change the controller interval while it's running or not. Otherwise, I'll try to do something more elegant :/

raybejjani committed:
endpoint: Sort slices in models.Endpoint

Consecutive calls to Endpoint.GetModel return slices with different
orders. By sorting them we simplify reflect.DeepCopy comparisions of
this type.

Signed-off-by: Ray Bejjani <ray@covalent.io>

tgraf commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -128,17 +148,57 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
+		CustomResourceDefinitionPluralName = "ciliumendpoints"
+
+		// CustomResourceDefinitionShortNames are the abbreviated names to refer to this CRD's instances
+		CustomResourceDefinitionShortNames = []string{"cep"}
```
`cep` was slightly confusing to me in the beginning as I was mapping it to `Endpoints` and not `Endpoint` but I can't think of anything better as `ce` is likely too short.

raybejjani replied:
yeah :/  I was following the cnp precedent (it also has ciliumnp but ciliumep seemed a little off). I could have `cep` and `ceps`.
I'm also open to other names, especially if they're more obviously not related to the k8s endpoints. On the other hand, we already call them endpoints so :(

aanm replied:
also, what about `ciliumep`? (Since we have `ciliumnp`)

raybejjani replied:
yah, sure. I'll add it.

tgraf commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -158,3 +159,50 @@ type CiliumNetworkPolicyList struct {
+	// +k8s:openapi-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	Status CiliumEndpointDetail `json:"status"`
```
I'm sure you thought about the current and future of this. Our current model is both `Spec` and `Status`. I'm OK to make it read-only and then add `Spec` later on and allowing to modify certain elements such as Configuration or Labels`

raybejjani replied:
This was part of the intention (aanm mentioned that k8s merged a change that allows using a status that cannot be edited by a user). It was, originally, `endpoint` but that wasn't too helpful. `status` isn't ideal, though, because the endpoint struct we use (the one also returned in `cilium endpoint get`) has a `status` field inside it.
Anyway, it should extend in a reasonable way.

tgraf replied:
I'm ok as-is

aanm replied:
Just FYI

>aanm mentioned that k8s merged a change that allows using a status that cannot be edited by a user

That was my understanding of that k8s merge regarding that feature. I was wrong. I've spent a couple hours trying to understand this behavior with k8s 1.11.

The `status` field in the CRD is writable by everyone. The `status` field on all other k8s objects is non-writable, in a sense k8s knows the `status` field should not be modified by external entities.

So, it's up to the controller of that CRD object, in our case cilium, to be aware of external changes of the `status` field and keep it updated as possible. We can still, and should, use `status` to report the `status` of a CRD.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -162,6 +174,30 @@ func init() {
+		ciliumEndpointSyncControllerK8sClient = k8sClient
+	})
+
+	if err != nil {
```
Does this really work? Won't the `:=` in the `func()` shadow `err`?

raybejjani replied:
nope, that totally doesn't work. I'll fix it. Thanks!

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -361,14 +397,122 @@ func (e *Endpoint) WaitForProxyCompletions() error {
+	// The health endpoint is created outside of k8s and shouldn't create a CEP
+	// for it. It has a specific label on it and we do not start the controller
+	// in that case.
+	healthLbl := labels.NewLabel(labels.IDNameHealth, "", labels.LabelSourceReserved)
```
What's the rationale to hide it?

tgraf replied:
We can consider showing it in the `kube-system` namespace

raybejjani replied:
I did have that to start (kube-system). It was hacky because we were forcing a k8s namespace into it but not any other k8s related field. I also worried that it would confuse people to see endpoints without any corresponding pods.
shrug. It's up to you, I don't really have a strong opinion.

tgraf replied:
I'm ok as-is

raybejjani replied:
oops, sorry, I put it back in yesterday. I can remove it, but I'll assume it doesn't matter that much and avoid the added churn.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -361,14 +397,122 @@ func (e *Endpoint) WaitForProxyCompletions() error {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has not changed")
+					return nil
+				}
+				lastMdl = mdl
```
It can still fail in which case we will never update again, I would update this last

raybejjani replied:
The sad part is that this check will never actually do anything. We include the controller status in mdl but this is a controller, so every iteration we see a different value from the last time. I don't know how to fix this without suppressing the controller information. That would undermine the point (to some extent). Another would be to ignore the status for this controller but it would depend on the frequency of other controllers then. What do you think?

tgraf replied:
You could store the last model in the endpoint

raybejjani replied:
Could you explain more what you're thinking? Obviously we can store the data but I'm not sure why it would be useful overall. Since we didn't have a `data` parameter on the DoFunc I assumed we intended to use the fact that you keep the state within the closure you create for DoFunc. As long as we're consistent, I like the idea of keeping controller state close to the functions themselves since it makes them far more self-contained (The only rebase conflicts I've had this PR have been imports, not the actual important code). This falls apart if they start editing fields on the endpoint willy nilly, however.

tgraf replied:
OK. The idea was to use the fact that we keep on passing the same endpoint pointer into updated controller functions. I initially had a state pointer in the controller but I observed that this means that we start keeping state copies around.

OTOH, we we assume that whoever owns the controller also controls the state then we can share state between controllers as long as locking is respected. For this particular case, I'm fine to add state to the controller itself to store the lastModel as it clearly only ever applies to the particular controller instance.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -361,14 +397,122 @@ func (e *Endpoint) WaitForProxyCompletions() error {
+				// do an update
+				case err == nil:
+					// Update the copy of the cep
+					(*cilium_v2.CiliumEndpointDetail)(mdl).DeepCopyInto(&cep.Status)
```
Should we check if `DeepCopyInto()` failed and left an empty cep?

raybejjani replied:
We should. The Deepcopy should never fail but there is other error indicator. +1

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -429,6 +573,10 @@ func NewEndpointFromChangeModel(base *models.EndpointChangeRequest, l pkgLabels.
 
+	ep.Mutex.Lock()
+	defer ep.Mutex.Unlock()
+	ep.runK8sCiliumEndpointSync()
```
I understand the desire of this but I think we need to move it somewhere else, in code like this:

``` go
        addLabels := labels.ParseStringLabels(lbls)
        ep, err := endpoint.NewEndpointFromChangeModel(epTemplate, addLabels)
        if err != nil {
                return PutEndpointIDInvalidCode, err
        }
        ep.SetDefaultOpts(d.conf.Opts)

        oldEp, err2 := endpointmanager.Lookup(id)
        if err2 != nil {
                return PutEndpointIDInvalidCode, err2
        } else if oldEp != nil {
                return PutEndpointIDExistsCode, fmt.Errorf("Endpoint ID %s exists", id)
        }
        if err = endpoint.APICanModify(ep); err != nil {
                return PutEndpointIDInvalidCode, err
        }

        if err := endpointmanager.AddEndpoint(d, ep, "Create endpoint from API PUT"); err != nil {
                log.WithError(err).Warn("Aborting endpoint join")
                return PutEndpointIDFailedCode, err
        }
```

it will cause controllers to be leaked in the error paths because they never get `Leave()` invoked on them.

How about we do it in `endpointmanager.AddEndpoint()` instead? It guarantees that we call `Leave()` again. I don't think we need  controller for any endpoint that is not managed by the endpointmanager.

raybejjani replied:
This is much cleaner. I'll try to do it (I hope there isn't an import cycle).

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -198,6 +199,81 @@ func getCiliumClient() (ciliumClient cilium_client_v2.CiliumV2Interface, err err
+//   - for each CEP
+//       delete CEP if the node does not exist
+// CiliumEndpoint objects follow a nodename-cep-endpointID scheme
+func RunK8sCiliumEndpointSyncGC() {
```
k8s 1.9 has added leases to cleanup the apiserver endpoints.
https://github.com/rphillips/kubernetes/commit/08966accfb40b6748b741613f7f7f40e8e24140e

Could we use it as well instead of GC?

tgraf replied:
I'm also fine to stick with GC for the initial version

raybejjani replied:
I'm still trying to find how they use the leases. It's really unclear to me if it's a more general concept or specific to this reconciler (unless you were thinking we could pretend each CRD is a "master" of itself). In particular, I'm not clear that anything is deleted vs the annotations acting as a priority ordering when picking a master.

tgraf replied:
I probably don't know more than you at this point. I saw this go in and observed that it addresses the same problem as we have in this case. Resources created by components.

aanm replied:
@raybejjani can you create a GH issue so we don't lose track of this?

raybejjani replied:
yup. https://github.com/cilium/cilium/issues/3101

tgraf reviewed (changes requested):
Awesome work overall. Some comments.

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -172,3 +173,50 @@ type CiliumNetworkPolicyList struct {
+
+// DeepCopyInto is an inefficient hack to allow reusing models.Endpoint in the
+// CiliumEndpoint CRD.
+func (in *CiliumEndpointDetail) DeepCopyInto(out *CiliumEndpointDetail) {
```
Does it hurt if we have the deepcopy generator in the `models` package to avoid this workaround? (I think it's doable)

raybejjani replied:
I think I can do that. I had left it here because DeepCopyInto is a k8s specific generated deepcopy, and i didn't want to confuse the types in models.

raybejjani replied:
Oh, no, they use the same thing. I'll change it to autogen in the other one.

raybejjani replied:
Hmm. Or not. I will look at this again later today. In the meantime, I'll push up my other fixups for review.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -163,6 +177,36 @@ func init() {
+		return nil, err
+	}
+
+	return ciliumEndpointSyncControllerK8sClient.CiliumV2(), nil
```
Is this intended? What happens if we are running k8s 1.6 and the cilium client should be `CiliumV1()`?

raybejjani replied:
We wouldn't have the CRD in the 1.6 case (I didn't create a TPR). I'll check the version and return an error, and that should avoid the impossible situation.

aanm replied:
That's true

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -434,6 +585,9 @@ func NewEndpointFromChangeModel(base *models.EndpointChangeRequest, l pkgLabels.
 		}
 	}
 
+	ep.Mutex.Lock()
```
What is this doing?

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -1001,6 +1155,9 @@ func ParseEndpoint(strEp string) (*Endpoint, error) {
 
 	ep.state = StateRestoring
 
+	ep.Mutex.Lock()
```
What is this doing?

raybejjani replied:
you know... locking things...
I think I had these when I started the sync controller in these functions (the start part needs the lock). I moved those and, clearly, didn't review my own code :/

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -500,6 +500,15 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
+	sort.StringSlice(mdl.Labels.Disabled).Sort()
+	sort.StringSlice(mdl.Labels.OrchestrationIdentity).Sort()
+	sort.StringSlice(mdl.Labels.OrchestrationInfo).Sort()
+	sort.Slice(mdl.Controllers, func(i, j int) bool { return mdl.Controllers[i].Name < mdl.Controllers[i].Name })
```
this will always return `false` :smile: (one is `i`, then the other one should be `j`)

raybejjani replied:
oops, yes... I mean, it kinda worked ;)

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -162,6 +174,16 @@ var (
+
+	// ciliumEPControllerLimit is the range of k8s versions with which we are
+	// willing to run the EndpointCRD controllers
+	ciliumEPControllerLimit, _ = go_version.NewConstraint(">= 1.7.0")
```
Please write `> 1.6` I've found out that constrain doesn't work with `1.7.0-alpha`

raybejjani replied:
done.

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -51,7 +42,7 @@ const (
 	// CustomResourceDefinitionSchemaVersion is semver-conformant version of CRD schema
 	// Used to determine if CRD needs to be updated in cluster
-	CustomResourceDefinitionSchemaVersion = "1.3"
+	CustomResourceDefinitionSchemaVersion = "1.4"
```
1.5

raybejjani replied:
done.

aanm reviewed (commented):
One last thing can you also update `./examples/kubernetes/rbac.yaml`?
